### DISCUSSION TITLE
ARC DPS module

### DISCUSSION THREAD

**Discussion by cedvan** [blish-hud/Blish-HUD#646] [Module Ideas]
Guild wars 2 does not support addons natively.
So the best way to bring new features is to use overlays like Blish HUD.

Use Addon Manager is therefore doomed to failure because no framework for addons is provided with the game. So Addon Manager to manage DLL is a bad way... Blish HUD on the other hand provides a magnificent framework for developing new modules. It also embeds a robust and powerful module management system.

Following these conclusions, Blish HUD has all the necessary qualities to become the main addons (modules) manager of Guild Wars 2.

But there is one detail that prevents him from going in this direction.: Historical addons.

- ARC DPS (+ extensions heal and mechanics): Sniffer combat stats, missing in GW2.
- d912pxy: GW2 recently added directx 11 and should allow improvements to handle multi core, so this addon should become useless. 
    - https://www.guildwars2.com/en/news/directx11-support-is-coming-to-guild-wars-2
    
> DirectX11 also offers some modern technology options that aren’t available in DirectX9. Upgrading to DirectX11 is the first step toward being able to do more shiny things.
After careful research we decided to integrate the open source rendering library BGFX into Guild Wars 2.

> We chose to use DirectX11 instead of DirectX12 or Vulkan because we found that switching to BGFX’s DirectX11 implementation provided enough of a performance boost that the graphics backend was no longer ever a limiting factor for client performance. DirectX11 is very stable and has already been used by thousands of games for nearly a decade at this point. It allows us to provide support back to Windows Vista, while Vulkan support starts at Windows 7 and DirectX12 support starts at Windows 10. 

- Radial: Blish HUD has a greater equivalent "Mounts" module
- Taco: Taco is an overlay and Blish HUD has a greater equivalent "Pathing" module

ARC DPS is therefore the only current addon that cannot be replaced by Blish HUD because the official GW2 API does not provide combat data. This is also why this addon is tolerated by ArenaNet.
On the other hand, Blish HUD is able to use the ARC DPS API to retrieve combat data and use it in its modules (example like Squad and Killproof modules).

To allow Blish HUD to become the main addon manager, it would be appropriate to integrate the installation and update of ARC DPS into the core of Blish HUD.

It would therefore be interesting to add:
- Automatic ARC DPS DLL download
- Automatic renaming to "d3d9.dll" or "d3d11.dll" depending on the engine used on GW2
- Copy DLL in game directory
- Same process to ARC DPS extensions with an activation from settings in Blish HUD
- Automatically added Blish ARC DPS bridge extension
- Add button to open ARC DPS settings window from Blish settings window
- Automatic deactivation of ARC DPS on crashes when starting the game
- Automatic check update to ARC DPS and download new DDL on starting Blinch, replace ARC DPS dll when game is not start.

Then any modules will be able to use the data from ARC DPS. And maybe a friendly user interface for arc DPS will emerge?

**Comment by coderedart** [comment]
This is not to replace Gw2 Addon Manager which handles management of arcdps and its addons along with other chainload dlls like reshade . 

The module requested here is just to 
1. install arcdps and the bridge into the game folder. 
2. restart the game if neccessary. 
3. update when available
4. allow an option to uninstall arcdps/bridge when its crashing the game. // either because game update or because latest arcdps has issues.

Although, the idea of arcdps addons as blish modules has merit. but there already exist a number of arcdps addons and they are not going to be rewritten for blish overnight. so, all of the above functionality must be optional and the present manual management of bridge/arcdps must also be continued to not interfere with existing users workflow.

this provides an alternative way of developing arcdps addons as blish modules instead. 

there's a few advantages:
1. blish modules don't crash the game if they mess up and are much easier to install / manage compared to the raw dlls
2. C# (or lua in nearby future) is much easier / convenient to develop with while providing reasonable performance compared to raw system languages like C/C++/rust
3. you get the whole framework of blish like UI or the Mono Engine to render stuff onto screen.

**Comment by dlamkins** [comment]
With upcoming changes to the ArcDPS bridge, we'll likely be investing more attention to it in Core.  As core can directly depend on the bridge, it is likely that we would support managing / installing the bridge in core - that specifically has been on the roadmap.

**Comment by Void2258** [comment]
Support for the installation and maintenance of arcs by Blish would be very helpful, even if it obviously cannot act as a normal Blish module due to underlying systemic issues as is. Arcs is the sole feature that necessitates the use of the Add On manager, as all other things it supports have Blish equivalents. It would be very handy if at some point Blish could take over these duties, allowing for an all-in-one add-on framework for distributing and keeping add-ons up to date. 
An eventual transition to a more standard module as detailed above would be nice, but is no doubt far out in the future if it happens at all.
